add: file processor(word counter) for web service

yoonsung0711-1 opened the pull request:
file processor 만들기 #6

(next) file processor 관련 테스트 코드 작성

yoonsung0711 committed:
add: file processor(word counter) for web service

NamsooCho commented on src/wc.rs:
```diff
@@ -0,0 +1,87 @@
+pub fn process() {
+    fn read_file(filename: &str) -> String {
+        let mut contents = String::new();
+        let mut file = File::open(filename).unwrap();
```
unwrap보다는 expect가 좋습니다.

yoonsung0711-1 replied:
다른 코드에서도 unwrap() 사용한 부분이 있는지 확인해서 expect로 변경하도록 하겠습니다.

> unwrap보다는 expect가 좋습니다.

NamsooCho commented on src/wc.rs:
```diff
@@ -0,0 +1,87 @@
+        let mut file = File::open(filename).unwrap();
+        let _ = file.read_to_string(&mut contents).unwrap();
+        let length = (&contents).len();
+        unsafe {
```
unsafe 안쓰고 해결 방법이 없나요? 있을 것 같은데요.

yoonsung0711-1 replied:
알려주신 것처럼 슬라이스에 split_at_mut라는 메소드가 있네요 ^^.. 해당 메소드로 교체하였습니다.

yoonsung0711-1 replied:
알려주신 것처럼 슬라이스에 split_at_mut라는 메소드가 있네요 ^^.. 해당 메소드로 교체하였습니다.

> unsafe 안쓰고 해결 방법이 없나요? 있을 것 같은데요.

NamsooCho commented on src/wc.rs:
```diff
@@ -0,0 +1,87 @@
+use std::collections::hash_map::RandomState;
+
+pub fn process() {
+    fn read_file(filename: &str) -> String {
```
네스팅된 함수가 좋은 패턴은 아닙니다.

yoonsung0711-1 replied:
말씀하신 것처럼 네스팅 시 테스트 함수 작성이 어려워져서
외부 함수로 빼내겠습니다. ^^..

> 네스팅된 함수가 좋은 패턴은 아닙니다.

NamsooCho commented on src/wc.rs:
```diff
@@ -0,0 +1,87 @@
+    let word_freq = frequencies(word_list);
+    // /*log/ */println!("{:?}", word_freq);
+
+    fn sort(word_freq: &HashMap<String, u16>) -> Vec<(&String, &u16)> {
```
sort를 직접 구현하는것 보다는 sort기능이 있는 컬렉션 사용이 좋습니다.

yoonsung0711-1 replied:
BTreeMap으로 변경해보도록 하겠습니다 ^^..

> sort를 직접 구현하는것 보다는 sort기능이 있는 컬렉션 사용이 좋습니다.

yoonsung0711-1 replied:
BTreeMap에 내장된 정렬 기능은 key 기준(word의 lexical order)이네요.. 
현재 라이브러리에서는 value(word frequency의 numerical order) 기준으로 sort하는 
메소드를 제공하고 있지 않아
기존 HashMap에서 sort를 구현한 것과 같은 방법을 쓸수 밖에 없을 것 같습니다. 

https://stackoverflow.com/questions/41220872/how-if-possible-to-sort-a-btreemap-by-value-in-rust

yoonsung0711-1 replied:
그럼에도 기존 코드는 BTreeMap을 사용해보기 위해
BTreeMap으로 변경하겠습니다. ^^..

NamsooCho replied:
이름을 데이터라 하든 키라하든, 정렬의 대상을 키로 잡아야 합니다. 데이터를 소팅 했으면 사실상 그 데이터가 키입니다.

NamsooCho replied:
자료 구조를 다시 생각해 보세요

yoonsung0711-1 replied:
단어 카운팅 시에는 단어의 유일성을 확보하면서 key의 중복 횟수를 세기 위해서
(key: 단어, value: 숫자)로 쌍을 맞춘 hashmap, hashset을 사용하고
카운팅이 다 끝난 이후에는 숫자가 가장 많은 순서로 정렬을 해서 상위 25개를 출력하도록 되어 있는데
단어와 달리 value는 유일성이 만족되지 않아(중복 가능성이 있어) 
(k, v) -> (v, k)로 바꾸는 건 여의치 않아 보입니다.

게다가 자료구조로 사용한 hashmap과 btreemap에는 value 기준으로 정렬하는 기능이 없어서
부득이 중간에 자료구조를 Vec<Pair<k, v>>로 변환한 뒤에 (정렬 후 순서 보전을 위해)
value 기준으로 정렬했는데..이 같은 사정은 다른 언어도 마찬가지라 kotlin으로 해당 기능을 구현할 때에도 비슷하게 처리 했었습니다. 

거듭 생각해도 hashmap을 래핑한 다른 자료구조를 만든다고 할지라도 
자료구조 변환과.. 정렬 연산은 사용하지 않는 방법이 떠오르지 않습니다 ^^..

NamsooCho commented on src/wc.rs:
```diff
@@ -0,0 +1,87 @@
+            println!("{:?} - {:?}", (*p.1).1, (*p.1).0);
+        }
+    }
+}
```
cr/lf

yoonsung0711-1 replied:
또 빼먹었네요 ^^.. 앞으로는 올리기 전에 확인하도록 하겠습니다.

> cr/lf

NamsooCho commented on pride-and-prejudice.txt:
```diff
@@ -0,0 +1,13426 @@
+﻿The Project Gutenberg EBook of Pride and Prejudice, by Jane Austen
```
data file 은 별도 디렉토리에 넣는게 좋아요.

yoonsung0711-1 replied:
resources 디렉토리를 만들도록 하겠습니다 ^^..

yoonsung0711 committed:
modify: file processor
(replace unwrap with expect,
take inner function out of closure,
add move static resources into sub directory)

NamsooCho commented on src/servers/single.rs:
```diff
@@ -45,17 +45,17 @@ pub fn single_thread_server(domain: &str, port_no: Option<u16>) -> Result<(), io
             println!("connection established!!!");
             stream
-        }).map_err(|error| { panic!("error!!! {:?}", error) }).unwrap();
+        }).map_err(|error| { panic!("error!!! {:?}", error) }).expect("connection failed");
```
panic과 expect를 둘다 쓸 필요가 없습니다. expect가 에러상황을 다 찍어주고 패닉 합니다.

NamsooCho commented on src/wc.rs:
```diff
@@ -0,0 +1,146 @@
+
+pub fn process() {
+    let contents = read_file("./resources/wc/pride-and-prejudice.txt");
+    // /*log/ */println!("{:?}", contents);
```
삭제하는 것이 나아 보입니다.

NamsooCho replied:
따로 로깅 크레이트를 쓰는 것도 방법이구요.

yoonsung0711-1 replied:
로그 크레이트 도입도 고려해보겠습니다 ^^.. 

> 따로 로깅 크레이트를 쓰는 것도 방법이구요.

NamsooCho commented on src/wc.rs:
```diff
@@ -0,0 +1,146 @@
+    return pair;
+}
+
+/* hashmap
```
프로젝트의 마스터 브랜치에 코멘트 처리된 코드를 머지하는 경우는 거의 없습니다.

yoonsung0711-1 replied:
코멘트랑 문서화 주석이랑 헷갈린 것 같습니다 ^^.. 로깅을 위해 주석 처리한 부분들도 logger로 대체하였습니다.

yoonsung0711 committed:
modify: replace console log with logger, remove unnecessary comments
